File: 0210_courseScheduleII/main.cpp
#include <vector>
#include <queue>

class Solution {
private:
    bool kahnAlgo(std::vector<std::vector<int>> &adj, int n, std::vector<int> &indegree, std::vector<int> &res)
    {
        std::queue<int> q;
        for (int i = 0; i < n; i++)
        {
            if (indegree[i] == 0)
                q.push(i);
        }

        int count = 0;
        while (!q.empty())
        {
            int curr = q.front();
            q.pop();

            for (auto a : adj[curr])
            {
                indegree[a] -= 1;
                if (indegree[a] == 0)
                    q.push(a);
            }
            res.push_back(curr);
            count++;
        }

        if (count != n)
            return false;
        return true;
    }
public:
    std::vector<int> findOrder(int numCourses, std::vector<std::vector<int>> &prerequisites)
    {
        int n = prerequisites.size();
        std::vector<std::vector<int>> adj(numCourses); // Adjacency matrix.
        std::vector<int> indegree(numCourses, 0); // indegree array.

        for (int i = 0; i < n; i++)
        {
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
            indegree[prerequisites[i][0]] += 1;
        }

        std::vector<int> res;
        if (kahnAlgo(adj, numCourses, indegree, res))
            return res;

        return {};
    }
};

File: 0210_courseScheduleII/main.py
from typing import List
import collections

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preq = {i:set() for i in range(numCourses)}
		# Create a graph for adjacency and traversing.
        graph = collections.defaultdict(set)
        for i,j in prerequisites:
            preq[i].add(j)
            graph[j].add(i)
        
        myQueue = collections.deque([])
        for k, v in preq.items():
            if len(v) == 0:
                myQueue.append(k)
        
        res = []
        while myQueue:
            course = myQueue.popleft()
            res.append(course)
            if len(res) == numCourses:
                return res
            for next in graph[course]:
                preq[next].remove(course)
                if not preq[next]:
                    myQueue.append(next)
        return []
    
    def findOrder1(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = collections.defaultdict(list)
        visited = set()
        recursion_visited = []
        res = []
        for i,j in prerequisites:
            graph[j].append(i)

        def dfs(i):
            visited.add(i)
            recursion_visited.append(i)
            for j in graph[i]:
                if j not in visited:
                    dfs(j)
                elif j in recursion_visited:
                    return
            res.append(i)
            recursion_visited.pop()    

        for i in range(numCourses):
            if i not in visited:
                dfs(i)
                
        return reversed(res) if len(res) == numCourses else []
    
